File: .agents/skills/discord-bot-engineer/assets/examples/discordjs/purge-progress/command.ts
import { MessageFlags, PermissionFlagsBits, type ChatInputCommandInteraction, type Message } from "discord.js";
import { createPurgeState, progressCopy, transition } from "./core.mjs";

export async function executePurge(interaction: ChatInputCommandInteraction, selected: Message[]): Promise<void> {
  if (!interaction.inCachedGuild() || !interaction.memberPermissions.has(PermissionFlagsBits.ManageMessages)) {
    await interaction.reply({ content: "คุณไม่มีสิทธิ์จัดการข้อความในห้องนี้", flags: MessageFlags.Ephemeral });
    return;
  }
  await interaction.deferReply({ flags: MessageFlags.Ephemeral });
  let state = transition(createPurgeState({ actorId: interaction.user.id, total: selected.length }), { type: "start" });
  await interaction.editReply(progressCopy(state));

  for (const message of selected) {
    let event;
    try {
      if (!message.deletable) event = { type: "batch", attempted: 1, deleted: 0, skipped: 1, failed: 0 };
      else {
        await message.delete();
        event = { type: "batch", attempted: 1, deleted: 1, skipped: 0, failed: 0 };
      }
    } catch {
      event = { type: "batch", attempted: 1, deleted: 0, skipped: 0, failed: 1 };
    }
    state = transition(state, event);
    if (state.attempted === state.total || state.attempted % 10 === 0) {
      await interaction.editReply(progressCopy(state));
    }
  }
  state = transition(state, { type: "finish" });
  await interaction.editReply(progressCopy(state));
}

File: .agents/skills/discord-bot-engineer/assets/examples/discordpy/purge_progress/command.py
from __future__ import annotations

import discord
from discord import app_commands

from .core import PurgeState, account, finish, progress_copy, start


@app_commands.command(name="purge", description="ลบข้อความพร้อมรายงานผลตามจริง")
@app_commands.checks.has_permissions(manage_messages=True)
async def purge(interaction: discord.Interaction, amount: app_commands.Range[int, 1, 100]) -> None:
    await interaction.response.defer(ephemeral=True)
    channel = interaction.channel
    if not isinstance(channel, discord.TextChannel):
        await interaction.edit_original_response(content="คำสั่งนี้ใช้ได้เฉพาะห้องข้อความ")
        return
    selected = [message async for message in channel.history(limit=amount)]
    state = start(PurgeState(actor_id=interaction.user.id, total=len(selected)))
    await interaction.edit_original_response(content=progress_copy(state))
    for message in selected:
        try:
            await message.delete()
            state = account(state, attempted=1, deleted=1, skipped=0, failed=0)
        except discord.Forbidden:
            state = account(state, attempted=1, deleted=0, skipped=1, failed=0)
        except discord.HTTPException:
            state = account(state, attempted=1, deleted=0, skipped=0, failed=1)
        if state.attempted == state.total or state.attempted % 10 == 0:
            await interaction.edit_original_response(content=progress_copy(state))
    await interaction.edit_original_response(content=progress_copy(finish(state)))
